# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
def add_two(x):
    return x +2
print(add_two(3))

5


In [2]:
lambda x: x + 2
(lambda x: x + 2)(3)

5

#### Saving to a Variable

In [3]:
f = lambda x, y: x+y+2
f(3,2)

7

#### Multiple Inputs

In [4]:
(lambda x, y, z: x*y*z)(3, 5, 8)




120

#### Passing a Lambda into a Function

In [5]:
def do_math(func, num):
    return func(num)

do_math(lambda x: x*x, 4)

16

#### Returning a Lambda from a Function

In [9]:
def returned_func():
    def squared(num):
        return num**2
    return squared

print(type(returned_func))
returned_func()()

def returne_func():
    return lambda num: num**2
print(returne_func()(3))

<class 'function'>
9


#### If Statements within Lambdas

In [12]:
#Lambda x : True if (condition) else false
f = lambda num: num*2 if num > 10 else num +2
f(8)

10

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [19]:
f = lambda x: x**3
f(5)

125

In [20]:
#get good at reading Documentation because this is your lifelinealso
help(sorted)

Help on built-in function sorted in module builtins:

sorted(iterable, /, *, key=None, reverse=False)
    Return a new list containing all items from the iterable in ascending order.
    
    A custom key function can be supplied to customize the sort order, and the
    reverse flag can be set to request the result in descending order.



In [28]:
#Using Lambdas with sort to sort a list of dictionary by a key from highest to smallest
from random import randint

my_class = [
    {"name": "Jody", "IQ": randint(1,240)},
    {"name": "Sarah", "IQ": randint(1,240)},
    {"name": "Dylan", "IQ": randint(1,240)},
    {"name": "Steven", "IQ": randint(1,240)}
]
print(my_class)
print(sorted(my_class, key= lambda student: student["IQ"], reverse = True))

[{'name': 'Jody', 'IQ': 169}, {'name': 'Sarah', 'IQ': 171}, {'name': 'Dylan', 'IQ': 44}, {'name': 'Steven', 'IQ': 59}]
[{'name': 'Sarah', 'IQ': 171}, {'name': 'Jody', 'IQ': 169}, {'name': 'Steven', 'IQ': 59}, {'name': 'Dylan', 'IQ': 44}]


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [29]:
# map(func,iterable(list,dict,tuple,set,etc...))
# Normally the usage of map happens with a pre-defined function -- but can use lambdas as well
help(map)

Help on class map in module builtins:

class map(object)
 |  map(func, *iterables) --> map object
 |  
 |  Make an iterator that computes the function using arguments from
 |  each of the iterables.  Stops when the shortest iterable is exhausted.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



In [31]:
def squared(num, num2):
    if num < 10 and num2 < 10:
        return num**2, num2**2
    else:
        return num, num2
numbers = [4,10,3,2,6]
more_nums = [4,11,20,3,15]

squared_nums_map = map(squared, numbers, more_nums)
list(squared_nums_map)


[(16, 16), (10, 11), (3, 20), (4, 9), (6, 15)]

In [32]:
squared_nums = list(map(lambda x: x**2 if x<10 else x,numbers))
print(squared_nums)

[16, 10, 9, 4, 36]


#### Using Lambda's with Map

In [34]:
# map(Lambda x+2,list)
# using Lambda in map happens in line (or in one line)
# list(map(lambda x: x*2,nums))
squared_nums = list(map(lambda x: x**2 if x<10 else x,numbers))
print(squared_nums)

[16, 10, 9, 4, 36]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [35]:
number= [4,10,9,8,7]
double = list(map(lambda x: x*2-1, numbers))
print(double)

[7, 19, 5, 3, 11]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [36]:
help(filter)

Help on class filter in module builtins:

class filter(object)
 |  filter(function or None, iterable) --> filter object
 |  
 |  Return an iterator yielding those items of iterable for which function(item)
 |  is true. If function is None, return the items that are true.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



In [103]:
list(filter(None, [True, True, False, False, "", "string"]))

[True, True, ' ', 'string']

#### Using Lambda's with Filter()

In [101]:
names= ["Shayne", "Gio", "Sarah", "Steven", "Fransico", "Dmitry"]
def s_names(name):
    if name[0].lower() == "s":
        return True
    else:
        return False
new_names_list = list(filter(s_names, names))
print(new_names_list)

['Shayne', 'Sarah', 'Steven']


In [102]:
new_names_lamb = list(filter(lambda name: True if name[0].lower()== '' else False, names))
new_names_lamb

[]

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [ ]:
nums =[2,7,4.2,1.6,9,4.4,4.9]



In [51]:
import statistics

nums =[2,7,4.2,1.6,9,4.4,4.9]

x = statistics.mean(nums)

below_avg = list(filter(lambda num: True if num < x else False, nums ))

print(x)
below_avg

4.728571428571429


[2, 4.2, 1.6, 4.4]

## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [52]:
from functools import reduce
# reduce(function,iterable)
help(reduce)



Help on built-in function reduce in module _functools:

reduce(...)
    reduce(function, sequence[, initial]) -> value
    
    Apply a function of two arguments cumulatively to the items of a sequence,
    from left to right, so as to reduce the sequence to a single value.
    For example, reduce(lambda x, y: x+y, [1, 2, 3, 4, 5]) calculates
    ((((1+2)+3)+4)+5).  If initial is present, it is placed before the items
    of the sequence in the calculation, and serves as a default when the
    sequence is empty.



In [54]:
list_1 = [2,4,6,8,10]
def subtract_nums(num1, num2):
    return num1 - num2
result= reduce(subtract_nums, list_1)
print(result)

-26


#### Using Lambda's with Reduce()

In [57]:
list_2 = [1, 10, 2, 30]
result_lamb = reduce(lambda num1, num2: num1 - num2, list_2)
print(result_lamb)


-41


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [59]:
l_5 = [1,2,3,4]

In [64]:
rl = reduce(lambda num1, num2: num1 * num2, l_5)
print(rl)

24


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

In [67]:
# this will break your shit.
#def recursive():
   # print("hi! ")
  #  recursive()
#recursive()

In [70]:
def recursive(i):
    if i == 4:
        return
    else:
        print("hi! ")
        i+=1
        recursive(i)
        
recursive(0)

hi! 
hi! 
hi! 
hi! 


#### Implementing a Base Case

In [186]:
def add_nums(num):
    if num <= 1:
        print("add_nums(1)=1")
        return num
    else:
        print(f'add_nums({num}) = {num} + add_nums({nums-1})')
        x = num - 1 
        return num + add_nums(x)
add_nums(5)

TypeError: unsupported operand type(s) for -: 'list' and 'int'

#### Writing a Factorial Function

In [78]:
# 5! = 5*4*3*2*1
def factorial(num):
    if num <= 1:
        print("factorial(1)=1")
        return 1
    else:
        print(f'factorial({num})={num} + factorial({num-1})')
        
factorial(10)


factorial(10)=10 + factorial(9)


#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [82]:
def sub_nums(num):
    if num <= 1:
        print("add_nums(1)=1")
        return num
    else:
        print(f'add_nums({num}) = {num} + add_nums({nums-1})')
        x = num - 1 
        return num + add_nums(x)
    
print(sub_nums())

TypeError: sub_nums() missing 1 required positional argument: 'num'

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

In [85]:
list(range(5))

[0, 1, 2, 3, 4]

In [86]:
def test():
    yield 1
    yield 2
    yield 3
    yield "a"
    yield ["a", "b", "c"]
    
for i in test():
    print(i)

1
2
3
a
['a', 'b', 'c']


#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [87]:
def my_range(n):
    for i in range(n):
        yield i
list(my_range(5))

[0, 1, 2, 3, 4]

In [88]:
def my_range(start, stop, step=1):
    while start<stop:
        yield start
        start += step
list(my_range(1,5,1))

[1, 2, 3, 4]

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [95]:
def range_squared(start, stop, step =1):
    while start > stop:
        yield start**2
        start -= step
        
for i in range_squared(10,0,1):
    print(i)

100
81
64
49
36
25
16
9
4
1


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]`

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [192]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]
new_places = list(filter(None, places))
print(new_places)
def place_string(places):
    for places in new_places:
        if ord(places[0]) != range(65, 90):
            return True
    else:
        return False
place = list(filter(place_string, new_places))
print(place)

[' ', 'Argentina', ' ', 'San Diego', '  ', 'Boston', 'New York']
[' ', 'Argentina', ' ', 'San Diego', '  ', 'Boston', 'New York']


In [153]:
import re
r=re.compile(r'[A-Za-z]+')
results=list(filter(r.match,places))
print(results)

['Argentina', 'San Diego', 'Boston', 'New York']


In [116]:
ord(",")

44

In [97]:
ord("A")

65

In [98]:
ord("Z")

90

### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [171]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]
author.sort(key=lambda name: name.split(" ")[-1].lower())
print(author)

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


In [162]:
help(sorted)

Help on built-in function sorted in module builtins:

sorted(iterable, /, *, key=None, reverse=False)
    Return a new list containing all items from the iterable in ascending order.
    
    A custom key function can be supplied to customize the sort order, and the
    reverse flag can be set to request the result in descending order.



### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]
`
`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [183]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

F = list(map(lambda place: (place[0], (9/5) * place[1] + 32), places))
print(F)

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


In [175]:
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]
print(type(places[0][1]))

<class 'int'>


### Exercise #4
Write a recursion function to perform the fibonacci sequence up to the number passed in.

Output for fib(5) => <br>
Iteration 0: 1 <br>
Iteration 1: 1 <br>
Iteration 2: 2<br>
Iteration 3: 3<br>
Iteration 4: 5<br>
Iteration 5: 8<br>

​

In [196]:
def fib(num):
    if num == 0:
        return 1
    elif num == 1:
        return 1
    else:
        return fib(num-2) + fib(num-1)


def fib_output(n):
    for i in range(n+1):
        print(f"Iteration {i}: {fib(i)}")
    

fib_output(5)

Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8
